File: SourceCode/lib/backend/media.dart
import 'dart:io';
import 'package:path/path.dart' as path;
import 'config.dart' show ConfigManager;

class MediaFile {
  final String name;
  final String filePath;
  final String fileType; // 'image' or 'video'
  final int size;
  final DateTime modified;

  MediaFile({
    required this.name,
    required this.filePath,
    required this.fileType,
    required this.size,
    required this.modified,
  });

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'path': filePath,
      'type': fileType,
      'size': size,
      'modified': modified.toIso8601String(),
    };
  }
}

class MediaManager {
  static String getAppDir() {
    if (Platform.isWindows) {
      final exePath = Platform.resolvedExecutable;
      return path.dirname(exePath);
    } else {
      return Directory.current.path;
    }
  }

  static String getMediaDir() {
    return path.join(getAppDir(), 'Media');
  }

  static Future<void> ensureMediaDir() async {
    final mediaDir = Directory(getMediaDir());
    if (!await mediaDir.exists()) {
      await mediaDir.create(recursive: true);
      print('✅ Created Media directory: ${getMediaDir()}');
    }
  }

  static Future<List<MediaFile>> getFiles() async {
    await ensureMediaDir();

    final mediaDir = Directory(getMediaDir());
    final files = <MediaFile>[];

    await for (final entity in mediaDir.list()) {
      if (entity is File) {
        final ext = path.extension(entity.path).toLowerCase();
        String? fileType;

        if (['.svg', '.png', '.jpg', '.jpeg'].contains(ext)) {
          fileType = 'image';
        } else if (ext == '.mp4') {
          fileType = 'video';
        }

        if (fileType != null) {
          final stat = await entity.stat();
          files.add(MediaFile(
            name: path.basename(entity.path),
            filePath: entity.path,
            fileType: fileType,
            size: stat.size,
            modified: stat.modified,
          ));
        }
      }
    }

    // Sort by name
    files.sort((a, b) => a.name.compareTo(b.name));

    print('📁 Found ${files.length} media files');
    return files;
  }

  static Future<void> saveFile(String filename, List<int> bytes) async {
    await ensureMediaDir();

    final filePath = path.join(getMediaDir(), filename);
    final file = File(filePath);

    await file.writeAsBytes(bytes);
    print('✅ Saved media file: $filename (${bytes.length} bytes)');
  }

  static Future<void> deleteFile(String filename) async {
    final filePath = path.join(getMediaDir(), filename);
    final file = File(filePath);

    if (!await file.exists()) {
      throw Exception('File not found: $filename');
    }

    // Security check: ensure file is in Media directory
    final canonicalPath = file.absolute.path;
    final canonicalMediaDir = Directory(getMediaDir()).absolute.path;

    if (!canonicalPath.startsWith(canonicalMediaDir)) {
      throw Exception('Invalid file path: $filename');
    }

    await file.delete();
    print('🗑️  Deleted media file: $filename');
  }

  static String getFilePath(String filename) {
    return path.join(getMediaDir(), filename);
  }
}

File: SourceCodeOLD-Tauri/src-tauri/src/media.rs
use serde::{Deserialize, Serialize};
use tokio::fs;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MediaFile {
    pub name: String,
    pub path: String,
    #[serde(rename = "type")]
    pub file_type: String,
    pub size: u64,
    pub modified: String,
}

pub fn get_media_dir() -> Result<std::path::PathBuf, String> {
    crate::paths::get_media_dir()
}

pub async fn get_files() -> Result<Vec<MediaFile>, String> {
    let media_dir = get_media_dir()?;
    
    if !media_dir.exists() {
        println!("DEBUG: Creating Media directory...");
        fs::create_dir_all(&media_dir).await
            .map_err(|e| format!("Failed to create Media directory: {}", e))?;
        println!("DEBUG: Media directory created");
        return Ok(Vec::new());
    }
    
    let mut entries = fs::read_dir(&media_dir).await
        .map_err(|e| format!("Failed to read Media directory: {}", e))?;
    let mut files = Vec::new();
    
    while let Some(entry) = entries.next_entry().await.map_err(|e| e.to_string())? {
        let path = entry.path();
        
        if path.is_dir() {
            continue;
        }
        
        if let Some(ext) = path.extension() {
            let ext_str = ext.to_string_lossy().to_lowercase();
            
            let file_type = match ext_str.as_str() {
                "svg" | "png" | "jpg" | "jpeg" => "image",
                "mp4" => "video",
                _ => continue,
            };
            
            let metadata = entry.metadata().await.map_err(|e| e.to_string())?;
            let modified = metadata.modified().map_err(|e| e.to_string())?;
            
            files.push(MediaFile {
                name: entry.file_name().to_string_lossy().to_string(),
                path: path.to_string_lossy().to_string(),
                file_type: file_type.to_string(),
                size: metadata.len(),
                modified: format!("{:?}", modified),
            });
        }
    }
    
    files.sort_by(|a, b| a.name.cmp(&b.name));
    
    println!("DEBUG: Found {} media files", files.len());
    
    Ok(files)
}

pub async fn delete_file(filename: &str) -> Result<(), String> {
    let media_dir = get_media_dir()?;
    let file_path = media_dir.join(filename);
    
    if !file_path.starts_with(&media_dir) {
        return Err("Invalid file path".to_string());
    }
    
    if !file_path.exists() {
        return Err("File not found".to_string());
    }
    
    fs::remove_file(&file_path).await
        .map_err(|e| e.to_string())?;
    
    Ok(())
}

pub async fn save_file(filename: &str, data: &[u8]) -> Result<(), String> {
    let media_dir = get_media_dir()?;
    let file_path = media_dir.join(filename);
    
    if !file_path.starts_with(&media_dir) {
        return Err("Invalid file path".to_string());
    }
    
    if !media_dir.exists() {
        fs::create_dir_all(&media_dir).await
            .map_err(|e| e.to_string())?;
    }
    
    fs::write(&file_path, data).await
        .map_err(|e| e.to_string())?;
    
    Ok(())
}